enterprise_access/apps/content_assignments/api.py

```diff
@@ -635,6 +640,51 @@ def _get_existing_assignments_for_allocation(
 
     return existing_assignments
 
+def _get_existing_assignments_for_requests(assignment_configuration, learner_credit_requests):
+    """
+    Finds all existing, re-allocatable assignments for a heterogeneous batch
+    of learner credit requests in a single, efficient query.
+
+    This correctly checks for matches on both (email, content_key) and
+    (lms_user_id, content_key).
+
+    Args:
+        assignment_configuration (AssignmentConfiguration): The configuration to search within.
+        learner_credit_requests (list[LearnerCreditRequest]): The list of requests.
+
+    Returns:
+        dict: A mapping of (learner_email, content_key) to the existing assignment object.
+    """
+    if not learner_credit_requests:
+        return {}
+
+    # Build a complex Q object to find all matches in one query.
+    # For each request, we look for an assignment that matches either the email/course
+    # combination OR the lms_user_id/course combination.
+    query = Q()
+    for request in learner_credit_requests:
+        # Always check for a match on the email and course key.
+        sub_query = Q(learner_email__iexact=request.user.email, content_key=request.course_id)
+
+        # If the request has a valid lms_user_id, also check for a match on that.
+        if request.user.lms_user_id:
+            sub_query |= Q(lms_user_id=request.user.lms_user_id, content_key=request.course_id)
```

Comment:
this is not needed as the a learner submitting a request will always be a valid user.